presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -0,0 +1,271 @@
+    }
+
+    private static class Visitor
+            extends io.prestosql.sql.tree.ExpressionRewriter<Void>
+    {
+        private final Metadata metadata;
+        private final TypeAnalyzer typeAnalyzer;
+        private final Session session;
+        private final TypeProvider types;
+        private final InterpretedFunctionInvoker functionInvoker;
+        private final LiteralEncoder literalEncoder;
+
+        public Visitor(Metadata metadata, TypeAnalyzer typeAnalyzer, Session session, TypeProvider types)
+        {
+            this.literalEncoder = new LiteralEncoder(metadata.getBlockEncodingSerde());
+            this.functionInvoker = new InterpretedFunctionInvoker(metadata.getFunctionRegistry());
+            this.metadata = metadata;
+            this.typeAnalyzer = typeAnalyzer;
+            this.session = session;
+            this.types = types;
+        }
+
+        @Override
+        public Expression rewriteComparisonExpression(ComparisonExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            ComparisonExpression expression = (ComparisonExpression) treeRewriter.defaultRewrite((Expression) node, null);
+            return unwrapCast(expression, typeAnalyzer, session, types, metadata, functionInvoker, literalEncoder);
+        }
+    }
+
+    private static Expression unwrapCast(ComparisonExpression expression, TypeAnalyzer typeAnalyzer, Session session, TypeProvider types, Metadata metadata, InterpretedFunctionInvoker functionInvoker, LiteralEncoder literalEncoder)
+    {
+        Cast cast;
+        Literal literal;
+        ComparisonExpression.Operator operator;
+
+        if (expression.getLeft() instanceof Cast && expression.getRight() instanceof Literal) {
+            cast = (Cast) expression.getLeft();
+            literal = (Literal) expression.getRight();
+            operator = expression.getOperator();
+        }
+        else if (expression.getLeft() instanceof Literal && expression.getRight() instanceof Cast) {
+            cast = (Cast) expression.getRight();
+            literal = (Literal) expression.getLeft();
+            operator = expression.getOperator().flip();
+        }
+        else {
+            return expression;
+        }
+
+        if (literal instanceof NullLiteral) {
+            return expression;
+        }
+
+        Type sourceType = typeAnalyzer.getType(session, types, cast.getExpression());
+        Type literalType = typeAnalyzer.getType(session, types, literal);
+
+        Optional<Type.Range> range = sourceType.getRange();
+        if (!range.isPresent()) {
+            return expression;
+        }
+
+        Object min = range.get().getMin();
+        Object max = range.get().getMax();
+
+        Object maxInTargetType = functionInvoker.invoke(
+                metadata.getFunctionRegistry().getCoercion(sourceType, literalType),
```

Comment:
I originally had a check to ensure there cast was an implicit cast (which implies the notion of "widening" conversion), but removed it thinking I might not need it. The example you gave is why we *do* need it :)

Reply:
@martint hm... can we know whether coercion _could be_ implicit? so that implicit coercion and identical-but-explicit coercion behave the same?
Also, implicit does not imply that it maintain distinctness. For example coercion of large integer values into `double` -- btw "saturated floor cast" suffers from the same problem and I remember we removed some such casts because they could produce results.
So, I think we might need something more explicit that implicit vs not implicit. Maybe a cast function could be annotated whether it maintains distinctness?

Reply:
I'm adding special checks for non-injective implicit coercions.